notify/notify.go

```diff
@@ -382,55 +382,47 @@ func (ws *WaitStage) Exec(ctx context.Context, alerts ...*types.Alert) (context.
 // DedupStage filters alerts.
 // Filtering happens based on a notification log.
 type DedupStage struct {
-	nflog nflog.Log
-	recv  *nflogpb.Receiver
+	nflog        nflog.Log
+	recv         *nflogpb.Receiver
+	sendResolved bool
 
-	// TODO(fabxc): consider creating an AlertBatch type received
-	// by stages that implements these functions.
-	// This can then also handle caching so we can skip passing
-	// the hash around as a context.
-	hash     func([]*types.Alert) []byte
-	resolved func([]*types.Alert) bool
-	now      func() time.Time
+	now  func() time.Time
+	hash func(*types.Alert) uint64
 }
 
 // NewDedupStage wraps a DedupStage that runs against the given notification log.
-func NewDedupStage(l nflog.Log, recv *nflogpb.Receiver) *DedupStage {
+func NewDedupStage(l nflog.Log, recv *nflogpb.Receiver, sendResolved bool) *DedupStage {
 	return &DedupStage{
-		nflog:    l,
-		recv:     recv,
-		hash:     hashAlerts,
-		resolved: allAlertsResolved,
-		now:      utcNow,
+		nflog:        l,
+		recv:         recv,
+		now:          utcNow,
+		sendResolved: sendResolved,
+		hash:         hashAlert,
 	}
 }
 
 func utcNow() time.Time {
 	return time.Now().UTC()
 }
 
-// TODO(fabxc): this could get slow, but is fine for now. We may want to
-// have something mor sophisticated at some point.
-// Alternatives are FNV64a as in fingerprints or xxhash.
-func hashAlerts(alerts []*types.Alert) []byte {
-	// The xor'd sum so we don't have to sort the alerts.
-	// XXX(fabxc): this approach caused collision issues with FNV64a in
-	// the past. However, sha256 should not suffer from the bit cancelation
-	// in in small input changes.
-	xsum := [sha256.Size]byte{}
+const sep = '\xff'
 
-	for _, a := range alerts {
-		b := make([]byte, 9)
-		binary.BigEndian.PutUint64(b, uint64(a.Fingerprint()))
-		// Resolved status is part of the identity.
-		if a.Resolved() {
-			b[8] = 1
-		}
-		for i, b := range sha256.Sum256(b) {
-			xsum[i] ^= b
-		}
+func hashAlert(a *types.Alert) uint64 {
+	b := make([]byte, 0, 1024)
```

Comment:
Can we get that one from a sync.Pool? Prometheus doesn't have it because mostly it does shortcut around the hash. (We probably still want to add it there eventually as well though as the QE hashes a lot.)